services/web/server/src/simcore_service_webserver/projects/db.py

```diff
@@ -444,6 +446,54 @@ async def get_project(
                 project_type,
             )
 
+    async def check_project_node_has_all_required_inputs(
+        self, user_id: UserID, project_uuid: ProjectID, node_id: NodeID
+    ) -> None:
+        project_dict, _ = await self.get_project(user_id, f"{project_uuid}")
+        workbench = project_dict["workbench"]
+
+        nodes_map: dict[str, Node] = {i: Node(**workbench[i]) for i in workbench}
+        node = nodes_map[f"{node_id}"]
+
+        unset_required_inputs: list[str] = []
+        unset_outputs_in_upstream: list[tuple[str, str]] = []
+
+        def _check_required_input(required_input_key: str) -> None:
+            input_entry: PortLink | None = None
+            if node.inputs:
+                input_entry = node.inputs.get(required_input_key, None)
+            if input_entry is None:
+                # NOT linked to any node connect service or set value manually(whichever applies)
+                unset_required_inputs.append(required_input_key)
+                return
+
+            source_node_id: str = f"{input_entry.node_uuid}"
+            source_output_key = input_entry.output
+
+            source_node = nodes_map[source_node_id]
+
+            output_entry: OutputsDict | None = None
+            if source_node.outputs:
+                output_entry = source_node.outputs.get(source_output_key, None)
+            if output_entry is None:
+                unset_outputs_in_upstream.append((source_output_key, source_node.label))
+
+        for required_input in node.inputs_required:
+            _check_required_input(required_input)
+
+        node_with_required_inputs = node.label
+        if unset_required_inputs:
+            msg = f"Missing '{', '.join(unset_required_inputs)}' connection(s) to '{node_with_required_inputs}'"
+            raise ProjectNodeRequiredInputsNotSetError(msg)
```

Comment:
I see now ... , you have to different message. 
Nonetheless, I recommend to have all messages as `msg_template` and use instead two different type of exceptions. We can talk offline about this.

Reply:
discussed offline